File: ur_cpp_control_ws/src/ur_cpp_control/src/imu_sim_node.cpp
#include <cmath>
#include <memory>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/imu.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Vector3.h"

namespace
{
constexpr double kDefaultFrequency = 0.2;
constexpr double kDefaultMagnitude = 0.5;
constexpr double kDefaultRate = 50.0;
}  // namespace

class ImuSimulator : public rclcpp::Node
{
public:
  ImuSimulator() : Node("imu_simulator"), start_time_(now())
  {
    frame_id_ = declare_parameter<std::string>("frame_id", "imu_link");
    publish_rate_hz_ = declare_parameter<double>("publish_rate_hz", kDefaultRate);
    oscillation_frequency_hz_ = declare_parameter<double>("oscillation_frequency_hz", kDefaultFrequency);
    oscillation_magnitude_rad_ = declare_parameter<double>("oscillation_magnitude_rad", kDefaultMagnitude);
    axis_ = declare_parameter<std::vector<double>>("oscillation_axis", {0.0, 0.0, 1.0});

    if (axis_.size() != 3) {
      RCLCPP_WARN(get_logger(), "oscillation_axis must have 3 elements. Using default axis.");
      axis_ = {0.0, 0.0, 1.0};
    }

    publisher_ = create_publisher<sensor_msgs::msg::Imu>("imu/data", rclcpp::SystemDefaultsQoS());

    const auto period = std::chrono::duration<double>(1.0 / publish_rate_hz_);
    timer_ = create_wall_timer(period, std::bind(&ImuSimulator::publish_imu, this));
  }

private:
  void publish_imu()
  {
    const auto now_time = now();
    const double t = (now_time - start_time_).seconds();
    const double angle = std::sin(2.0 * M_PI * oscillation_frequency_hz_ * t) *
                         oscillation_magnitude_rad_;

    tf2::Vector3 axis(axis_[0], axis_[1], axis_[2]);
    if (axis.length2() < 1e-6) {
      axis = tf2::Vector3(0.0, 0.0, 1.0);
    }
    axis.normalize();

    tf2::Quaternion q(axis, angle);
    q.normalize();

    sensor_msgs::msg::Imu msg;
    msg.header.stamp = now_time;
    msg.header.frame_id = frame_id_;
    msg.orientation.x = q.x();
    msg.orientation.y = q.y();
    msg.orientation.z = q.z();
    msg.orientation.w = q.w();

    publisher_->publish(msg);
  }

  rclcpp::Time start_time_;
  std::string frame_id_;
  double publish_rate_hz_;
  double oscillation_frequency_hz_;
  double oscillation_magnitude_rad_;
  std::vector<double> axis_;

  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImuSimulator>());
  rclcpp::shutdown();
  return 0;
}

File: ur_imu_teleop_ws/src/ur_imu_teleop/ur_imu_teleop/imu_sim_node.py
import math

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu


def quaternion_from_euler(roll: float, pitch: float, yaw: float):
    cr = math.cos(roll * 0.5)
    sr = math.sin(roll * 0.5)
    cp = math.cos(pitch * 0.5)
    sp = math.sin(pitch * 0.5)
    cy = math.cos(yaw * 0.5)
    sy = math.sin(yaw * 0.5)
    w = cr * cp * cy + sr * sp * sy
    x = sr * cp * cy - cr * sp * sy
    y = cr * sp * cy + sr * cp * sy
    z = cr * cp * sy - sr * sp * cy
    return x, y, z, w


class ImuSimulator(Node):
    def __init__(self):
        super().__init__("imu_simulator")
        self.declare_parameter("frame_id", "imu_link")
        self.declare_parameter("publish_rate_hz", 50.0)
        self.declare_parameter("roll_amplitude_rad", 0.6)
        self.declare_parameter("pitch_amplitude_rad", 0.4)
        self.declare_parameter("yaw_amplitude_rad", 0.8)
        self.declare_parameter("angular_speed_rad_s", 0.6)

        self.publisher = self.create_publisher(Imu, "imu/data", 10)
        rate = self.get_parameter("publish_rate_hz").get_parameter_value().double_value
        self.timer = self.create_timer(1.0 / rate, self.publish_imu)
        self.start_time = self.get_clock().now()

    def publish_imu(self):
        now = self.get_clock().now()
        t = (now - self.start_time).nanoseconds * 1e-9
        roll_amp = self.get_parameter("roll_amplitude_rad").get_parameter_value().double_value
        pitch_amp = self.get_parameter("pitch_amplitude_rad").get_parameter_value().double_value
        yaw_amp = self.get_parameter("yaw_amplitude_rad").get_parameter_value().double_value
        omega = self.get_parameter("angular_speed_rad_s").get_parameter_value().double_value

        roll = roll_amp * math.sin(omega * t)
        pitch = pitch_amp * math.sin(omega * 0.7 * t)
        yaw = yaw_amp * math.sin(omega * 1.3 * t)
        qx, qy, qz, qw = quaternion_from_euler(roll, pitch, yaw)

        msg = Imu()
        msg.header.stamp = now.to_msg()
        msg.header.frame_id = self.get_parameter("frame_id").get_parameter_value().string_value
        msg.orientation.x = qx
        msg.orientation.y = qy
        msg.orientation.z = qz
        msg.orientation.w = qw
        msg.angular_velocity.x = 0.0
        msg.angular_velocity.y = 0.0
        msg.angular_velocity.z = 0.0
        msg.linear_acceleration.x = 0.0
        msg.linear_acceleration.y = 0.0
        msg.linear_acceleration.z = 0.0
        self.publisher.publish(msg)


def main():
    rclpy.init()
    node = ImuSimulator()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
